### PR TITLE
Implement serialize, update tests

### PR DESCRIPTION
**Pull request by ptous**
Updates to the latest pyo3, implements `Serialize` and updates tests to account for it.

This is a pretty small change set, but I'm a relatively new Rust developer so it's possible I'm doing silly things. Feedback quite welcome.

This has seen active use in one of my own projects, adapting [this project](https://gitlab.com/rubidium-dev/powers) to serialize its output into Python dicts. Originally I approached this with [dict-derive](https://github.com/gperinazzo/dict-derive), but serde's control over field attributes and serialization paths was much more effective.  Without that level of control, repeated reference counted fields were being re-serialized to Python many thousands of times. I could have created stripped-down copies of the structs I was after and serialized those instead, but using serde eliminates the need to maintain two sets of structs as things evolve.

I didn't _need_ to implement `Deserialize` so I haven't yet, but I _could_ use it in my project, so I might yet take a stab. If I do, I'll PR that as well.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -153,8 +159,8 @@ impl<'py> Serializer for Pythonizer<'py> {
     fn serialize_f64(self, v: f64) -> Result<Self::Ok, Self::Error> {
         Ok(v.into_py(self.py))
     }
-    fn serialize_char(self, _v: char) -> Result<Self::Ok, Self::Error> {
-        todo!()
+    fn serialize_char(self, v: char) -> Result<Self::Ok, Self::Error> {
+        Ok(v.to_string().into_py(self.py))
```
Comment (ptous) [inline src/lib.rs:163]: I went for a `str` representation here since this is a unicode character and not just bytes.

File: src/lib.rs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 ///
 /// ```
 /// use serde::{Serialize, Deserialize};
-/// use pyo3::{Python, AsPyRef};
+/// use pyo3::{Python};
```
Comment (davidhewitt) [inline src/lib.rs:5]:
```suggestion
/// use pyo3::Python;
```

File: src/lib.rs [hunk 3]
```diff
@@ -42,17 +45,20 @@ pub fn depythonize<T>(_py: Python, _obj: PyObject) -> T {
...
-    fn custom<T>(_msg: T) -> Self
+    fn custom<T>(msg: T) -> Self
     where
         T: std::fmt::Display,
     {
-        todo!()
+        let s: String = format!("{}", msg);
+        PythonizerError {
+            0: PyErr::new::<PyException, _>(s)
+        }
```
Comment (davidhewitt) [inline src/lib.rs:55]: `std::fmt::Display` gives us a blanket impl of `std::string::ToString`. Also I prefer the `PyException::new_err` constructors personally:

```suggestion
        PythonizerError(PyException::new_err(s.to_string()))
```

File: src/lib.rs [hunk 4]
```diff
@@ -42,17 +45,20 @@ pub fn depythonize<T>(_py: Python, _obj: PyObject) -> T {
...
     }
 }
 
 impl std::fmt::Display for PythonizerError {
-    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        todo!()
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self.0)
```
Comment (davidhewitt) [inline src/lib.rs:61]: I would err on the side of using `Display` instead of `Debug` in a `Display` impl:

```suggestion
        write!(f, "{}", self.0)
```

File: src/lib.rs (hunk 1 above)
Comment (davidhewitt) [reply src/lib.rs:163]: That sounds reasonable. One other possible design would be to serialize the same as `u32` and output a scalar value. I wasn't sure what would be more useful for users, so happy to listen to your opinion on this.

(On the python side the code could then use the `chr()` builtin if they want to convert the scalar value to a `str`.)

File: src/lib.rs [hunk 5]
```diff
@@ -331,9 +337,10 @@ impl ser::SerializeStruct for PythonDictSerializer<'_> {
     where
         T: Serialize,
     {
+        let py = PyNativeType::py(self.dict);
         Ok(self
             .dict
-            .set_item(key, pythonize(self.dict.py(), value)?)?)
+            .set_item(key, pythonize(py, value)?)?)
```
Comment (davidhewitt) [inline src/lib.rs:343]: Curious why this change was needed here? `dict.py()` is just a more concise form of `PyNativeType::py(self.dict)`

**Review by davidhewitt — APPROVED** [review]
Thanks, this is great stuff and very much appreciated! 

I have a few minor points, once they're addressed I'll merge this, create a CHANGELOG and release this as version 0.2.

**Comment by davidhewitt** [comment]
Thanks again for this PR, I've made the suggested changes and merging now. 🚀
